drivers: sensors: bmi08x: add initial support for bmi08x

XenuIsWatching opened the pull request:
This adds support for the bosch bmi085 and bmi088. This also includes support for data sync mode. The binary blob in bmi08x_config_file.h is under the BSD-3 clause which comes from the github repo https://github.com/BoschSensortec/BMI08x-Sensor-API/blob/master/LICENSE 

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

carlescufi commented:
I want to note that the binary blob for configuration is something we already allow in similar drivers:

- https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/bmi270/bmi270_config_file.h

My recommendation is to document that this is allowed.

MaureenHelm commented on drivers/sensor/bmi08x/Kconfig:
```diff
@@ -0,0 +1,226 @@
+	bool "250 DPS"
+
+config BMI08X_GYRO_RANGE_125DPS
+	bool "125 DPS"
+endchoice
+
+choice BMI08X_GYRO_ODR
+	prompt "Gyroscope sampling frequency."
+	default BMI08X_GYRO_ODR_2000_532 if BMI08X_DATA_SYNC_ODR_2000
+	default BMI08X_GYRO_ODR_1000_116 if BMI08X_DATA_SYNC_ODR_1000
+	default BMI08X_GYRO_ODR_400_47 if BMI08X_DATA_SYNC_ODR_400
+	default BMI08X_GYRO_ODR_RUNTIME
+
+config BMI08X_GYRO_ODR_RUNTIME
+	bool "Set at runtime." if !BMI08X_DATA_SYNC
+
+config BMI08X_GYRO_ODR_100_12
+	bool "100 Hz, 12Hz filter" if !BMI08X_DATA_SYNC
+
+config BMI08X_GYRO_ODR_100_32
+	bool "100 Hz, 32Hz filter" if !BMI08X_DATA_SYNC
+
+config BMI08X_GYRO_ODR_200_23
+	bool "200 Hz, 23Hz filter" if !BMI08X_DATA_SYNC
+
+config BMI08X_GYRO_ODR_200_64
+	bool "200 Hz, 64Hz filter" if !BMI08X_DATA_SYNC
+
+config BMI08X_GYRO_ODR_400_47
+	bool "400 Hz, 47Hz filter" if ((BMI08X_DATA_SYNC && BMI08X_DATA_SYNC_ODR_400) || !BMI08X_DATA_SYNC)
+
+config BMI08X_GYRO_ODR_1000_116
+	bool "1000 Hz, 116Hz filter" if ((BMI08X_DATA_SYNC && BMI08X_DATA_SYNC_ODR_1000) || !BMI08X_DATA_SYNC)
+
+config BMI08X_GYRO_ODR_2000_230
+	bool "2000 Hz, 230Hz filter" if ((BMI08X_DATA_SYNC && BMI08X_DATA_SYNC_ODR_2000) || !BMI08X_DATA_SYNC)
+
+config BMI08X_GYRO_ODR_2000_532
+	bool "2000 Hz, 532Hz filter" if ((BMI08X_DATA_SYNC && BMI08X_DATA_SYNC_ODR_2000) || !BMI08X_DATA_SYNC)
+endchoice
```
Move these to devicetree properties to allow per-instance configuration

XenuIsWatching replied:
not fixed yet in the last push, but will get to

XenuIsWatching replied:
done

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,781 @@
+		return -EIO;
+	}
+
+	k_busy_wait(100);
```
Can we sleep instead of busy-wait?

XenuIsWatching replied:
fixed

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_config_file.h:
```diff
@@ -0,0 +1,426 @@
+/* Bosch BMI08X inertial measurement unit header
+ *
+ * Copyright (c) 2022 Meta Platforms, Inc. and its affiliates
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
The source copyright and license should be preserved (Bosch, BSD-3-Clause)

XenuIsWatching replied:
fixed.... but it appears to of broken one of the CI checks

XenuIsWatching commented on dts/bindings/sensor/bosch,bmi08x-accel-i2c.yaml:
```diff
@@ -0,0 +1,51 @@
+        Zephyr driver maps all interrupts to INT. The signal defaults
+        to output low when produced by the sensor.
+
+    int1-map-io:
+      type: int
+      required: false
+      description: |
+        Enables INT1 as an output
+
+    int2-map-io:
+      type: int
+      required: false
+      description: |
+        Enables INT2 as an output
```
these are actually mapping either data ready, fifo watermark, or fifo full interrupts. I'll add a better description to it

XenuIsWatching commented on drivers/sensor/bmi08x/Kconfig:
```diff
@@ -0,0 +1,226 @@
+
+menuconfig BMI08X
+	bool "Bosch BMI08X inertial measurement unit"
+	depends on SPI || I2C
```
fixed

XenuIsWatching commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,781 @@
+		return -EIO;
+	}
+	/* required when switching power modes */
+	k_busy_wait(USEC_PER_MSEC * BMI08X_POWER_CONFIG_DELAY);
```
fixed

XenuIsWatching commented on drivers/sensor/bmi08x/bmi08x_gyro_trigger.c:
```diff
@@ -0,0 +1,137 @@
+
+#include <drivers/gpio.h>
+#include <drivers/sensor.h>
+#include <kernel.h>
```
fixed

XenuIsWatching commented on dts/bindings/sensor/bosch,bmi08x-accel-spi.yaml:
```diff
@@ -0,0 +1,51 @@
+include: [sensor-device.yaml, spi-device.yaml]
+
+properties:
+    int-gpios:
```
fixed

XenuIsWatching committed:
drivers: sensors: bmi08x: add initial support for bmi08x

This adds support for the bosch bmi085 and bmi088. This also includes
support for data sync mode.

Signed-off-by: Ryan McClelland <ryanmcclelland@meta.com>

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel_trigger.c:
```diff
@@ -0,0 +1,153 @@
+#endif
+
+	/* set accel ints */
+	if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
```
I didn't find it in the datasheet of bmi088 and bmi085 BMI08X_REG_ACCEL_INT1_MAP: 0x56 and BMI08X_REG_ACCEL_INT2_MAP: 0x57. the datasheet said that 0x55-0x57 is reserved.

XenuIsWatching replied:
it is there.... Just about every register that says "reserved" in that data actually does have a purpose (just not a very important one...) if you look in this Bosch public BMI08x bare metal driver.. you see those registers actually defined in there

Cherish-Gww replied:
> 

I really don`t like "reserved" register, I can`t find those "reserved" register describe.

XenuIsWatching replied:
it's here:
https://github.com/boschsensortec/BMI08x-Sensor-API/blob/631a966891885643f180e18e94870411b8f4c3fc/bmi08a.c#L2225

Cherish-Gww replied:
> 

I think this is not data ready interrupt,  data ready interrupt setting is `set_accel_data_ready_int`
by writing BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA register.

XenuIsWatching replied:
I see... I'll replace it with this
```c
#if BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
	if (config->data_sync != 0) {
		/* set accel ints */
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT2_MAP, cfg->int2_map) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
	} else
#endif
	{
		uint8_t map_data = ((cfg->int2_map << BMI08X_ACCEL_INT2_DRDY_POS) |
			(cfg->int1_map << BMI08X_ACCEL_INT1_DRDY_POS));
		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA, map_data) < 0) {
			LOG_ERR("Failed to map interrupts.");
			return -EIO;
		}
	}
```

Cherish-Gww replied:
I think it would be better to set the sensor synchronization function in the attribute. What do you think?
Macros can only be selected from one of two options, and using attribute functions can be set at any time.

> I see, I assume you're not using data sync?... I'll replace it with this
> 
> ```c
> #if BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
> 	if (config->data_sync != 0) {
> 		/* set accel ints */
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_MAP, cfg->int1_map) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT2_MAP, cfg->int2_map) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 	} else
> #endif
> 	{
> 		uint8_t map_data = ((cfg->int2_map << BMI08X_ACCEL_INT2_DRDY_POS) |
> 			(cfg->int1_map << BMI08X_ACCEL_INT1_DRDY_POS));
> 		if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_INT1_INT2_MAP_DATA, map_data) < 0) {
> 			LOG_ERR("Failed to map interrupts.");
> 			return -EIO;
> 		}
> 	}
> ```

I think it would be better to set the sensor synchronization function in the attribute. What do you think?
Macros can only be selected from one of two options, and using attribute functions can be set at any time.

XenuIsWatching replied:
Sorry, ...but I can't tell exactly what you're saying here??

Data sync is done in the DTS so it is instanced. You can have multiple BMI08X connected and one can have data sync on while the other doesn't. With a KConfig, it's either ALL or NONE.

Cherish-Gww replied:
> ```c
> BMI08X_ACCEL_ANY_INST_HAS_DATA_SYNC
> ```

`bmi08x_attr_set`

XenuIsWatching replied:
I emailed a bosch FAE about this a while ago... data sync can't be set a runtime

XenuIsWatching commented:
Other than the Kconfig CI issue... I'm thinking about how to bring `DATA_SYNC` to the DTS. While a bool would be the simplest, I'd also need to add build_asserts as there is some dependencies on the odr frequencies of both of the bmi08x-accel and bmi08x-gyro, otherwise it won't work.

I'm think it be easiest to use a phandle for the data-sync definition which would point to each other in the DTS, where those parameters in the DTS can be read from each other. If something is there (provided it is the other bmi08x-accel/gyro), then it would be implied that data-sync is enabled.

Cherish-Gww commented on drivers/sensor/bmi08x/Kconfig:
```diff
@@ -0,0 +1,98 @@
+menuconfig BMI08X
+	bool "Bosch BMI08X inertial measurement unit"
+	default y
+	depends on DT_HAS_BOSCH_BMI08X_ENABLED
```
depends on DT_HAS_BOSCH_BMI08X_ACCEL_ENABLED ||  DT_HAS_BOSCH_BMI08X_GYRO_ENABLED

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel_trigger.c:
```diff
@@ -0,0 +1,153 @@
+#define DT_DRV_COMPAT bosch_bmi08x_accel
+#include "bmi08x.h"
+
+#include <logging/log.h>
```
#include <zephyr/logging/log.h>

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,525 @@
+	{50, 0}, {100, 0}, {200, 0}, {400, 0}, {800, 0}, {1600, 0}, {3200, 0},
+};
+
+static int bmi08x_freq_to_odr_val(uint16_t freq_int, uint16_t freq_milli)
```
Prompt for repeated definition at compile time
in bmi08x_accel.c it has a same function name

Cherish-Gww replied:
I think this part is public and should be separated to bmi08x.c

XenuIsWatching replied:
yeah, and there's some other functions around it that ought to be common between the two. I'll get to moving it

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,525 @@
+static int bmi08x_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bmi08x_gyro_data *bmi08x = dev->data;
```
struct bmi08x_gyro_data *bmi08x = dev->data;
no use variable

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,525 @@
+
+int bmi08x_gyro_init(const struct device *dev)
+{
+	const struct bmi08x_gyro_config *config = dev->config(dev);
```
const struct bmi08x_gyro_config *config = dev->config;

dev->config is not a function call

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x.h:
```diff
@@ -0,0 +1,570 @@
+	const struct bmi08x_gyro_bus_io *api;
+#if defined(CONFIG_BMI08X_GYRO_TRIGGER)
+	struct gpio_dt_spec int_gpio;
+#endif
```
lost 
	uint8_t gyro_hz;
	uint8_t gyro_fs;

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,772 @@
+static int bmi08x_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bmi08x_accel_data *data = dev->data;
```
struct bmi08x_accel_data *data = dev->data;
no use variable

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,772 @@
+			      struct sensor_value *val)
+{
+	struct bmi08x_accel_data *data = dev->data;
+	uint8_t temp_raw[2] = {0};
```
uint8_t temp_raw[2] = {0};
no use variable

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,772 @@
+		LOG_ERR("Cannot write index");
+		return -EIO;
+	}
+	/* write config file */
```
Please add 
#ifdef CONFIG_BMI08X_DATA_SYNC
.....
#endif

or bmi08x_config_file will no found

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel_trigger.c:
```diff
@@ -0,0 +1,154 @@
+
+	gpio_init_callback(&bmi08x->gpio_cb, bmi08x_acc_gpio_callback, BIT(cfg->int_gpio.pin));
+
+	int ret = gpio_add_callback(bmi08x->gpio, &bmi08x->gpio_cb);
```
bmi08x->gpio is NULL, use cfg->int_gpio.port

Cherish-Gww replied:
> 

And Please delete the accel device data  gpio, no need it.

XenuIsWatching replied:
done

Cherish-Gww replied:
> done

@XenuIsWatching  I found that gyro init:
`	/* reboot the chip */
	if (bmi08x_gyro_byte_write(dev, BMI08X_REG_GYRO_SOFTRESET, BMI08X_SOFT_RESET_CMD) < 0) {
        }`
will failed on I2C bus, do you know why?

XenuIsWatching replied:
What do you mean by "failed"? what error code? This is working for me

Cherish-Gww replied:
> What do you mean by "failed"? what error code? This is working for me

-5, it means -EIO.

XenuIsWatching replied:
Also, please start discussion in a separate thread about this

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro_trigger.c:
```diff
@@ -0,0 +1,135 @@
+
+	gpio_init_callback(&bmi08x->gpio_cb, bmi08x_gyr_gpio_callback, BIT(cfg->int_gpio.pin));
+
+	int ret = gpio_add_callback(bmi08x->gpio, &bmi08x->gpio_cb);
```
bmi08x->gpio is NULL, use cfg->int_gpio.port

Cherish-Gww replied:
> bmi08x->gpio is NULL, use cfg->int_gpio.port

And Please delete the accel device data  gpio, no need it.

XenuIsWatching replied:
done

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,775 @@
+
+	pm_device_busy_set(dev);
+
+	if (bmi08x_accel_transceive(dev, BMI08X_REG_ACCEL_X_LSB | (1 << 7), false, data->acc_sample,
```
BMI08X_REG_ACCEL_X_LSB | (1 << 7)

ACC_X_LSB address is 0x12, I do not know why (0x12 | (1 << 7)) = 0x92?

Cherish-Gww replied:
In transceive_spi you can add reg address with  (uint8_t addr = reg & 0x7f for write uint8_t addr = reg | 0x80 for read.)

XenuIsWatching replied:
I should be using the BIT(7) macro here, but that bit is used to indicate a read

Cherish-Gww replied:
In datasheet
```
6.1.2 SPI interface of accelerometer part
The data bits are used as follows: 
• Bit #0: Read/Write bit. When 0, the data SDI is written into the chip. When 1, the data SDO from 
the chip is read. 
• Bit #1-7: Address AD(6:0). 
• Bit #8-15:  
o When in write mode, these are the data SDI, which will be written into the address. 
o When in read mode, these bits contain unpredictable values, and the user has to read 
Bit #16-23 to get the actual data from the reading address.  
```
that means only in spi mode read/write is using the register address Bit0, 
```
6.2 Inter-Integrated Circuit (I²C)
I²C write access and I²C read access
```
in I2C mode the read/write is in slave address bit7

the operate in register address or slave address will be set in low level api, not here.

XenuIsWatching replied:
if you read the rest of the code, and further into the zephyr i2c api, you'll see that r/w bit is handled there...

that "bit 0" (using the datasheet terminology) of the reg address is just a dummy in I2C, so if bit 0 is still set, it doesn't matter

Cherish-Gww replied:
> if you read the rest of the code, and further into the zephyr i2c api, you'll see that r/w bit is handled there...
> 
> that "bit 0" (using the datasheet terminology) of the reg address is just a dummy in I2C, so if bit 0 is still set, it doesn't matter

I am using bmi088 in I2C mode, in this function`if (bmi08x_accel_transceive(dev, BMI08X_REG_ACCEL_X_LSB | (1 << 7), false, data->acc_sample,`   it read the register address is 0x92 not 0x12, so I could not get right value

XenuIsWatching replied:
What value are you getting? That MSb is a dummy bit.

Are you getting the correct chip id as well? I'd be reading register 0x80 for that.

Cherish-Gww replied:
I see. you have changed it, this commit can be closed
```
	if (bmi08x_accel_read(dev, BMI08X_REG_ACCEL_X_LSB, (uint8_t *)data->acc_sample,
			      sizeof(data->acc_sample)) < 0) {
		pm_device_busy_clear(dev);
		return -EIO;
	}
```

and
```
	if (bmi08x_gyro_read(dev, BMI08X_REG_GYRO_X_LSB, (uint8_t *)bmi08x->gyr_sample,
			     sizeof(bmi08x->gyr_sample)) < 0) {
		return -EIO;
	}
```

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,472 @@
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	if (bmi08x_gyro_transceive(dev, BMI08X_REG_GYRO_X_LSB | (1 << 7), false, bmi08x->gyr_sample,
```
BMI08X_REG_GYRO_X_LSB | (1 << 7)

RATE_X_LSB address is 0x02, and I do not know why (0x02 | (1 << 7)) = 0x82?

Cherish-Gww replied:
In transceive_spi you can add reg address with  (uint8_t addr = reg & 0x7f for write uint8_t addr = reg | 0x80 for read.)

XenuIsWatching replied:
I can definitely tell English is not your first language here, but I'm not quite sure what you're trying to say here

bit 7 is to indicate is for a read, I probably could make a more clear by adding a macro
```
#define REG_READ_BIT BIT(7)
```

Cherish-Gww replied:
> I can definitely tell English is not your first language here, but I'm not quite sure what you're trying to say here
> 
> bit 7 is to indicate is for a read, I probably could make a more clear by adding a macro
> 
> ```c
> #define REG_READ_BIT BIT(7)
> ```

what I means is operate register address or slave address must in the low level api not here.

XenuIsWatching replied:
Your comment still doesn't make sense... but taking a look at this again, I should just replace this with the `bmi08x_gyro_read` function instead

Cherish-Gww replied:
> Your comment still doesn't make sense to me and there is no requirement for it to be in the low level api... but taking a look at this again, I should just replace this with the `bmi08x_gyro_read` function instead

Let me put it this way:
In I2C mode, control read/write bit is in slave address bit7.
In SPI mode, control read/write bit is in register address bit0.

slave address is different with register address. `BMI08X_REG_GYRO_X_LSB` is a register address.

XenuIsWatching replied:
You still haven't reviewed I2C yet... please do so.

XenuIsWatching replied:
The 7b I2C address is always the same.... I know you're talking about the rw bit, but that is handled by the i2c API

Cherish-Gww replied:
I see. you have changed it, this commit can be closed
```
	if (bmi08x_accel_read(dev, BMI08X_REG_ACCEL_X_LSB, (uint8_t *)data->acc_sample,
			      sizeof(data->acc_sample)) < 0) {
		pm_device_busy_clear(dev);
		return -EIO;
	}
```

and
```
	if (bmi08x_gyro_read(dev, BMI08X_REG_GYRO_X_LSB, (uint8_t *)bmi08x->gyr_sample,
			     sizeof(bmi08x->gyr_sample)) < 0) {
		return -EIO;
	}
```

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,775 @@
+
+int bmi08x_accel_read(const struct device *dev, uint8_t reg_addr, uint8_t *data, uint8_t len)
+{
+	return bmi08x_accel_transceive(dev, reg_addr | BIT(7), false, data, len);
```
reg_addr | BIT(7)

In spi or i2c the address operate is different, I think reg address can not operate here.

XenuIsWatching replied:
I do not understand what you are trying to say here, but BIT(7) Must be set to indicate a read of the reg

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,775 @@
+			   (.int_gpio = GPIO_DT_SPEC_INST_GET(inst, int_gpios),))                 \
+			BMI08X_ACCEL_TRIGGER_PINS(inst)                                            \
+				.accel_hz = DT_INST_ENUM_IDX(inst, accel_hz) + 5,                  \
+		.accel_fs = DT_INST_ENUM_IDX(inst, accel_fs) / 2, BMI08X_DATA_SYNC_REG(inst)};     \
```
.accel_fs = DT_INST_ENUM_IDX(inst, accel_fs) / 2

this is a bug, you need change to DT_INST_PROP(inst, accel_fs)
because in `bmi08x_acc_range_set(dev, config->accel_fs);` the range parameter need the real rang not the index. 

same as gyro sensor fs setting

XenuIsWatching replied:
Nope, the enum idx is to correspond with the reg value, please take a look at the datasheet again, the divide by 2 is there because the bmi085 and bmi088 have different settings

Cherish-Gww replied:
> 
first, I have tested this is mistake. I tried to explain why it was a mistake
In function `static int bmi08x_acc_range_set(const struct device *dev, int32_t range)`, called in `bmi08x_accel_init` to setting accel sensor range, the parameter range is using to get register value by `reg_val = bmi08x_range_to_reg_val(range, bmi085_acc_range_map, BMI085_ACC_RANGE_MAP_SIZE);`

and in function `bmi08x_range_to_reg_val`
```
int32_t bmi08x_range_to_reg_val(uint16_t range, const struct bmi08x_range *range_map,
				       uint16_t range_map_size)
{
	int i;

	for (i = 0; i < range_map_size; i++) {
		if (range <= range_map[i].range) {
			return range_map[i].reg_val;
		}
	}

	return -EINVAL;
}```

the range parameter is compare with 
```
static const struct bmi08x_range bmi085_acc_range_map[] = {
	{2, BMI085_ACCEL_RANGE_2G},
	{4, BMI085_ACCEL_RANGE_4G},
	{8, BMI085_ACCEL_RANGE_8G},
	{16, BMI085_ACCEL_RANGE_16G},
};
```
you can see that range map value is `2, 4, 8, 16`, this is range value not index.

XenuIsWatching replied:
I see, sorry, I was confused with another definition in my head. I will fix

Cherish-Gww replied:
> I see, sorry, I was confused with another definition in my head. I will fix

same as gyro sensor range setting, and `static const struct bmi08x_range bmi08x_gyr_range_map[] = {` need sort from small to large.

XenuIsWatching replied:
> > I see, sorry, I was confused with another definition in my head. I will fix
> 
> same as gyro sensor range setting, and `static const struct bmi08x_range bmi08x_gyr_range_map[] = {` need sort from small to large.

done

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,472 @@
+		IF_ENABLED(CONFIG_BMI08X_GYRO_TRIGGER,                                             \
+			   (.int_gpio = GPIO_DT_SPEC_INST_GET(inst, int_gpios),))                 \
+			.gyro_hz = DT_INST_ENUM_IDX(inst, gyro_hz),                                \
+		BMI08X_GYRO_TRIGGER_PINS(inst).gyro_fs = DT_INST_ENUM_IDX(inst, gyro_fs),          \
```
`gyro_fs = DT_INST_ENUM_IDX(inst, gyro_fs)`
is a bug, you need change it to `.gyro_fs = DT_INST_PROP(inst, gyro_fs),`
get range from dts, not the index.

XenuIsWatching replied:
Nope, the enum idx is to correspond with the reg value, please take a look at the datasheet again

Cherish-Gww replied:
the function `bmi08x_range_to_reg_val` need range value not index, and not the register value.

Cherish-Gww commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,472 @@
+	return bmi08x_gyro_byte_write(dev, reg_addr, (old_val & ~mask) | ((val << pos) & mask));
+}
+
+static const struct bmi08x_range bmi08x_gyr_range_map[] = {
```
The sorting here should be from small to large

XenuIsWatching replied:
No, the reg value goes from large to small, please take look at reg 0x0F: GYRO_RANGE in the datasheet

Cherish-Gww commented:
@XenuIsWatching 
A final suggestion: I2C and SPI can be divided into two independent files according to the interface, 
such as: https://github.com/zephyrproject-rtos/zephyr/tree/main/drivers/sensor/bme280

MaureenHelm commented on drivers/sensor/bmi08x/Kconfig:
```diff
@@ -0,0 +1,93 @@
+	bool "Bosch BMI08X inertial measurement unit"
+	default y
+	depends on DT_HAS_BOSCH_BMI08X_ACCEL_ENABLED || DT_HAS_BOSCH_BMI08X_GYRO_ENABLED
+	select I2C if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI08X_ACCEL),i2c) || $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI08X_GYRO),i2c)
+	select SPI if $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI08X_ACCEL),spi) || $(dt_compat_on_bus,$(DT_COMPAT_BOSCH_BMI08X_GYRO),spi)
```
Not sure why this didn't get flagged by the compliance check, but will you please shorten these long lines?

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,775 @@
+	struct bmi08x_accel_data *data = dev->data;
+	size_t i;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```
return an errno

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,775 @@
+		.accel_fs = DT_INST_PROP(inst, accel_fs), BMI08X_DATA_SYNC_REG(inst)};             \
+                                                                                                   \
+	PM_DEVICE_DT_INST_DEFINE(inst, bmi08x_accel_pm_action);                                    \
+	DEVICE_DT_INST_DEFINE(inst, bmi08x_accel_init, PM_DEVICE_DT_INST_GET(inst),                \
```
Please use `SENSOR_DEVICE_DT_INST_DEFINE`

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,472 @@
+	struct bmi08x_gyro_data *bmi08x = dev->data;
+	size_t i;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```
return an errno

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_gyro.c:
```diff
@@ -0,0 +1,472 @@
+	};                                                                                         \
+                                                                                                   \
+	PM_DEVICE_DT_INST_DEFINE(inst, bmi08x_gyro_pm_action);                                     \
+	DEVICE_DT_INST_DEFINE(inst, bmi08x_gyro_init, PM_DEVICE_DT_INST_GET(inst),                 \
```
Use `SENSOR_DEVICE_DT_INST_DEFINE`

MaureenHelm reviewed (changes requested):
A few minor nits, but otherwise looks good. The TSC approved #52652 so we can override the scancode check once all other code review comments have been resolved.

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_accel_trigger.c:
```diff
@@ -0,0 +1,168 @@
+static void bmi08x_handle_drdy_acc(const struct device *dev)
+{
+	struct bmi08x_accel_data *bmi08x = dev->data;
+	struct sensor_trigger drdy_trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+	};
```
Store a reference to the user-supplied trigger in `bmi08x_trigger_set_acc` and pass it back when invoking the handler. This will allow the application to use CONTAINER_OF with the trigger

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_gyro_trigger.c:
```diff
@@ -0,0 +1,136 @@
+static void bmi08x_handle_drdy_gyr(const struct device *dev)
+{
+	struct bmi08x_gyro_data *bmi08x = dev->data;
+	struct sensor_trigger drdy_trigger = {
+		.type = SENSOR_TRIG_DATA_READY,
+	};
```
Store a reference to the user-supplied trigger in bmi08x_trigger_set_gyr and pass it back when invoking the handler. This will allow the application to use CONTAINER_OF with the trigger

XenuIsWatching replied:
can you point me to an example of this being done?

MaureenHelm commented on drivers/sensor/bmi08x/bmi08x_gyro_trigger.c:
```diff
@@ -0,0 +1,136 @@
+	struct bmi08x_gyro_data *bmi08x = dev->data;
+	const struct bmi08x_gyro_config *cfg = dev->config;
+
+	if (!device_is_ready(cfg->int_gpio.port)) {
```
gpio_is_ready_dt

MaureenHelm reviewed (changes requested):
Needs rebase

MaureenHelm commented:
Please rebase to pick up a fix for the failed Pull Request Assigner check

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_own.conf:
```diff
@@ -44,5 +44,5 @@ CONFIG_STTS751_TRIGGER_OWN_THREAD=y
 CONFIG_VCNL4040_TRIGGER_OWN_THREAD=y
-CONFIG_WSEN_HIDS_TRIGGER_OWN_THREAD=y
-CONFIG_WSEN_TIDS_TRIGGER_OWN_THREAD=y
+CONFIG_BMI08X_ACCEL_TRIGGER_OWN_THREAD=y
+CONFIG_BMI08X_GYRO_TRIGGER_OWN_THREAD=y
```
Don't remove the WSEN drivers.

Please also add this new driver to `sensors_trigger_none.conf` and `sensors_trigger_global.conf`

XenuIsWatching replied:
yeah, that was likely an issue I made when rebasing and fixing merge conflicts >_<

MaureenHelm commented on tests/drivers/build_all/sensor/sensors_trigger_global.conf:
```diff
@@ -49,3 +49,5 @@ CONFIG_TMP007_TRIGGER_GLOBAL_THREAD=y
 CONFIG_VCNL4040_TRIGGER_GLOBAL_THREAD=y
 CONFIG_WSEN_HIDS_TRIGGER_GLOBAL_THREAD=y
 CONFIG_WSEN_TIDS_TRIGGER_GLOBAL_THREAD=y
+CONFIG_BMI08X_ACCEL_TRIGGER_GLOBAL_THREAD=y
+CONFIG_BMI08X_GYRO_TRIGGER_GLOBAL_THREAD=y
```
Move up to keep alpha order. Please do this for the other conf files too

MaureenHelm reviewed (changes requested):
Needs rebase

bperseghetti commented:
@MaureenHelm @XenuIsWatching is this able to be finally merged?

MaureenHelm commented:
> @MaureenHelm @XenuIsWatching is this able to be finally merged?

This PR needs one additional approval and then it will be ready to be merged. It will also need to wait for the merge window to reopen after the upcoming 3.4 release is complete.
https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management

bperseghetti commented:
@danieldegrasse are you able to review this? Or does it need a review from one of the listed reviewers?

carlescufi reviewed (dismissed):
~The scancode issues need to be fixed~

XenuIsWatching commented:
> @XenuIsWatching please rebase

rebased, but it blew away the reviews

bperseghetti commented:
@MaureenHelm @carlescufi sorry to bug you guys again on this, but are you able to re-approve and merge now that it's rebased?

bbilas commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,778 @@
+
+static int bmi08x_bus_check_i2c(const union bmi08x_bus *bus)
+{
+	return device_is_ready(bus->i2c.bus) ? 0 : -ENODEV;
```
```suggestion
	return i2c_is_ready_dt(&bus->i2c) ? 0 : -ENODEV;
```

bbilas commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,778 @@
+		reg_val = bmi08x_range_to_reg_val(range, bmi088_acc_range_map,
+						  BMI088_ACC_RANGE_MAP_SIZE);
+	} else {
+		return -ENXIO;
```
`ENODEV` is more common.

bbilas commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,778 @@
+		return -ENOTSUP;
+	}
+
+	return 0;
```
It seems to be superfluous.

bbilas commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,778 @@
+
+	if (bmi08x_accel_byte_write(dev, BMI08X_REG_ACCEL_PWR_CONF, conf_reg_val) < 0) {
+		LOG_ERR("Failed to set conf power mode");
+		return -EIO;
```
Is there any particular reason why you are not propagating the return codes from `_read/_write` functions?

XenuIsWatching replied:
I honstely can't remember why by this point... I completed this code 2 years ago and only just opened a PR in October...

I fixed it in both the accel and gyro code

bperseghetti commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -740,8 +740,30 @@ test_i2c_tcs3400: tcs3400@72 {
 
-test_tcn75a: tcn75a@73 {
+test_tcn75a: tcn75a@74 {
 	compatible = "microchip,tcn75a";
-	reg = <0x73>;
+	reg = <0x74>;
```
I believe this maybe got messed up in the rebase? I think it should be 73 not 74 from this [commit](https://github.com/zephyrproject-rtos/zephyr/commit/d3828418c09b2f35463da841a893798bc4ab9ce7): https://github.com/zephyrproject-rtos/zephyr/blob/fed377f622c9c280016c41b22ea6dc6b8ce0aed4/tests/drivers/build_all/sensor/i2c.dtsi#L743-L745

bperseghetti commented:
@XenuIsWatching thanks for the fixes! Can't wait wait to get this merged in!

carlescufi reviewed (dismissed):
refreshing my +1

XenuIsWatching commented on drivers/sensor/bmi08x/bmi08x_accel.c:
```diff
@@ -0,0 +1,808 @@
+			len1 = CONFIG_BMI08X_I2C_WRITE_BURST_SIZE;
+		}
+		if (bmi08x_stream_transfer_write_i2c(dev, index, data, len1) != 0) {
+			return -EIO;
```
actually 3 more... I took a second look through it all

carlescufi reviewed (approved):
Refresh +1

bperseghetti commented:
Thanks @bbilas! Are you able to merge this in now @carlescufi?

carlescufi commented:
> Thanks @bbilas! Are you able to merge this in now @carlescufi?

@MaureenHelm is back, so I delegate to her now.
